lib/Sema/SemaBounds.cpp

```diff
@@ -2088,9 +2015,37 @@ namespace {
                                       BoundsExpr *DeclaredBounds, Expr *Src,
                                       BoundsExpr *SrcBounds,
                                       bool InCheckedScope) {
+      // Record expression equality implied by initialization.
+      SmallVector<SmallVector <Expr *, 4> *, 4> EquivExprs;
+      SmallVector<Expr *, 4> EqualExpr;
+      if (S.CheckIsNonModifying(Src, Sema::NonModifyingContext::NMC_Unknown, false)) {
+        // TODO: make sure variable being initialized isn't read by Src.
+        DeclRefExpr *TargetDeclRef =
+          DeclRefExpr::Create(S.getASTContext(), NestedNameSpecifierLoc(),
+                              SourceLocation(), D, false, SourceLocation(),
+                              D->getType(), ExprValueKind::VK_LValue);
+        CastKind Kind;
+        QualType TargetTy;
+        if (D->getType()->isArrayType()) {
+          Kind = CK_ArrayToPointerDecay;
+          TargetTy = S.getASTContext().getArrayDecayedType(D->getType());
+        } else {
+          Kind = CK_LValueToRValue;
+          TargetTy = D->getType();
+        }
+        Expr *TargetExpr = BoundsInference(S).CreateImplicitCast(TargetTy, Kind, TargetDeclRef);
+        EqualExpr.push_back(TargetExpr);
+        EqualExpr.push_back(Src);
+        EquivExprs.push_back(&EqualExpr);
+        /*
```

Comment:
Maybe remove this commented-out block or un-comment it and guard it with `Trace` or something?

Reply:
Yes, I think this information should be available under a flag.    I'll do that in a future PR because I think we want to put some effort into presenting the information.

We have a challenge when the compiler can't prove something true or false of letting the programmer understand what was going on.  When I was looking at the results of bounds declaration checking, for situations where there was a warning, it was helpful to know what equality facts the compiler knew.   I think we want an explanatory mode that tells the programmer what facts the compiler knows and what it tried to prove.   The compiler could issue additional notes including this information.